storage: TestSingleKey failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/ffb539f6451a23fdcf8a50ad96a1166d77fa565d

Parameters:
```
TAGS=
GOFLAGS=-race
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=843778&tab=buildLog

```
=== RUN   TestSingleKey
W180817 07:07:35.534752 148416 server/status/runtime.go:294  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180817 07:07:36.715346 148416 server/server.go:828  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180817 07:07:36.716101 148416 base/addr_validation.go:260  [n?] server certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I180817 07:07:36.716184 148416 base/addr_validation.go:300  [n?] web UI certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I180817 07:07:36.954661 148416 server/config.go:496  [n?] 1 storage engine initialized
I180817 07:07:36.954979 148416 server/config.go:499  [n?] RocksDB cache size: 128 MiB
I180817 07:07:36.955085 148416 server/config.go:499  [n?] store 0: in-memory, size 0 B
I180817 07:07:37.967647 148416 server/node.go:373  [n?] **** cluster bd32192e-e9d2-439d-be56-5148d23940b9 has been created
I180817 07:07:37.968050 148416 server/server.go:1399  [n?] **** add additional nodes by specifying --join=127.0.0.1:42079
I180817 07:07:37.970005 148416 gossip/gossip.go:383  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:42079" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:11 > build_tag:"v2.1.0-alpha.20180702-1701-gffb539f" started_at:1534489657969557679 
I180817 07:07:38.642915 148416 storage/store.go:1506  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180817 07:07:38.784425 148416 server/node.go:476  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=6.9 KiB), ranges=1, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=7103.00 p25=7103.00 p50=7103.00 p75=7103.00 p90=7103.00 pMax=7103.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I180817 07:07:38.785116 148416 storage/stores.go:242  [n1] read 0 node addresses from persistent storage
I180817 07:07:38.786351 148416 server/node.go:697  [n1] connecting to gossip network to verify cluster ID...
I180817 07:07:38.786674 148416 server/node.go:722  [n1] node connected via gossip and verified as part of cluster "bd32192e-e9d2-439d-be56-5148d23940b9"
I180817 07:07:38.787162 148416 server/node.go:546  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I180817 07:07:38.884654 148416 server/status/recorder.go:652  [n1] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180817 07:07:38.884960 148416 server/server.go:1805  [n1] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180817 07:07:38.931185 148728 server/status/recorder.go:652  [n1,summaries] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
W180817 07:07:38.976231 148518 storage/closedts/provider/provider.go:143  [ct-closer] unable to move closed timestamp forward: node not in the liveness table
I180817 07:07:38.978802 148416 server/server.go:1536  [n1] starting https server at 127.0.0.1:34563 (use: 127.0.0.1:34563)
I180817 07:07:38.979069 148416 server/server.go:1538  [n1] starting grpc/postgres server at 127.0.0.1:42079
I180817 07:07:38.979152 148416 server/server.go:1539  [n1] advertising CockroachDB node at 127.0.0.1:42079
W180817 07:07:39.068056 148416 jobs/registry.go:308  [n1] unable to get node liveness: node not in the liveness table
I180817 07:07:39.454156 148462 storage/replica_command.go:282  [split,n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /System/"" [r2]
I180817 07:07:41.316976 148496 storage/replica_command.go:282  [split,n1,s1,r2/1:/{System/-Max}] initiating a split of this range at key /System/NodeLiveness [r3]
I180817 07:07:41.856130 148774 rpc/nodedialer/nodedialer.go:92  [consistencyChecker,n1,s1,r1/1:/{Min-System/}] connection to n1 established
I180817 07:07:42.238481 148772 sql/event_log.go:126  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
W180817 07:07:45.849885 148670 storage/store.go:3784  [n1,s1] handle raft ready: 0.9s [processed=1]
W180817 07:07:46.270184 148735 storage/node_liveness.go:558  [n1,hb] slow heartbeat took 2.8s
I180817 07:07:46.765321 148836 storage/replica_command.go:282  [split,n1,s1,r3/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/NodeLivenessMax [r4]
I180817 07:07:48.518993 148639 storage/replica_consistency.go:129  [consistencyChecker,n1,s1,r2/1:/System/{-NodeLive…}] triggering stats recomputation to resolve delta of {ContainsEstimates:true LastUpdateNanos:1534489663217380346 IntentAge:0 GCBytesAge:0 LiveBytes:-22070 LiveCount:-461 KeyBytes:-21468 KeyCount:-461 ValBytes:-602 ValCount:-461 IntentBytes:0 IntentCount:0 SysBytes:0 SysCount:0}
I180817 07:07:48.900605 148726 server/status/runtime.go:433  [n1] runtime stats: 1.3 GiB RSS, 266 goroutines, 20 MiB/104 MiB/146 MiB GO alloc/idle/total, 40 MiB/83 MiB CGO alloc/total, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (1580x)
I180817 07:07:49.234390 148558 storage/replica_command.go:282  [split,n1,s1,r4/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/tsd [r5]
I180817 07:07:50.103510 148728 server/status/recorder.go:652  [n1,summaries] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
W180817 07:07:51.753091 148826 storage/intent_resolver.go:668  [n1,s1] failed to push during intent resolution: failed to push "split" id=d0fb7276 key=/Local/Range/System/NodeLivenessMax/RangeDescriptor rw=true pri=0.00864696 iso=SERIALIZABLE stat=PENDING epo=0 ts=1534489669.234625106,0 orig=1534489669.234625106,0 max=1534489669.234625106,0 wto=false rop=false seq=1
I180817 07:07:52.232962 148821 sql/event_log.go:126  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:$1 User:root}
I180817 07:07:52.967356 148845 storage/replica_command.go:282  [split,n1,s1,r5/1:/{System/tsd-Max}] initiating a split of this range at key /System/"tse" [r6]
I180817 07:07:55.691156 148853 storage/replica_command.go:282  [split,n1,s1,r6/1:/{System/tse-Max}] initiating a split of this range at key /Table/SystemConfigSpan/Start [r7]
I180817 07:07:56.427030 148648 storage/replica_proposal.go:214  [n1,s1,r6/1:/{System/tse-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=2 start=1534489658.135072753,0 exp=1534489676.835931231,0 pro=1534489667.836057401,0
I180817 07:07:56.466115 148650 storage/replica_proposal.go:214  [n1,s1,r4/1:/System/{NodeLive…-tsd}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.456495440,0 following repl=(n1,s1):1 seq=2 start=1534489658.135072753,0 exp=1534489676.835931231,0 pro=1534489667.836057401,0
I180817 07:07:56.556781 148653 storage/replica_proposal.go:214  [n1,s1,r5/1:/System/ts{d-e}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.524765946,0 following repl=(n1,s1):1 seq=2 start=1534489658.135072753,0 exp=1534489676.835931231,0 pro=1534489667.836057401,0
I180817 07:07:57.780580 148830 sql/event_log.go:126  [n1,intExec=disableNetTrace] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:root}
I180817 07:07:58.496364 148670 storage/replica_proposal.go:214  [n1,s1,r6/1:/{System/tse-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:07:58.604732 148862 storage/replica_command.go:282  [split,n1,s1,r7/1:/{Table/System…-Max}] initiating a split of this range at key /Table/11 [r8]
I180817 07:07:59.013247 148726 server/status/runtime.go:433  [n1] runtime stats: 1.3 GiB RSS, 268 goroutines, 14 MiB/108 MiB/146 MiB GO alloc/idle/total, 40 MiB/84 MiB CGO alloc/total, 356.18cgo/sec, 0.23/0.01 %(u/s)time, 0.02 %gc (2x)
I180817 07:08:00.102405 148728 server/status/recorder.go:652  [n1,summaries] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180817 07:08:00.439186 148522 storage/replica_proposal.go:214  [n1,s1,r7/1:/{Table/System…-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:00.524522 148963 storage/replica_command.go:282  [split,n1,s1,r8/1:/{Table/11-Max}] initiating a split of this range at key /Table/12 [r9]
I180817 07:08:02.019741 148666 storage/replica_proposal.go:214  [n1,s1,r8/1:/{Table/11-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:02.125425 148980 storage/replica_command.go:282  [split,n1,s1,r9/1:/{Table/12-Max}] initiating a split of this range at key /Table/13 [r10]
I180817 07:08:02.606736 148966 sql/event_log.go:126  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:gen_random_uuid()::STRING User:root}
I180817 07:08:03.417188 148686 storage/replica_proposal.go:214  [n1,s1,r9/1:/{Table/12-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
W180817 07:08:03.588203 148686 storage/store.go:3784  [n1,s1] handle raft ready: 0.5s [processed=1]
I180817 07:08:03.605746 148958 storage/replica_command.go:282  [split,n1,s1,r10/1:/{Table/13-Max}] initiating a split of this range at key /Table/14 [r11]
I180817 07:08:04.856452 148957 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I180817 07:08:05.653247 148667 storage/replica_proposal.go:214  [n1,s1,r10/1:/{Table/13-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:05.845385 148940 storage/replica_command.go:282  [split,n1,s1,r11/1:/{Table/14-Max}] initiating a split of this range at key /Table/15 [r12]
W180817 07:08:06.323685 148666 storage/store.go:3784  [n1,s1] handle raft ready: 0.6s [processed=1]
I180817 07:08:06.426528 148876 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
W180817 07:08:07.002406 149029 server/server.go:1362  The server appears to be unable to contact the other nodes in the cluster. Please try

- starting the other nodes, if you haven't already
- double-checking that the '--join' and '--listen'/'--advertise' flags are set up correctly
- running the 'cockroach init' command if you are trying to initialize a new cluster

If problems persist, please see https://www.cockroachlabs.com/docs/v2.1/cluster-setup-troubleshooting.html.
I180817 07:08:07.131813 148521 storage/replica_proposal.go:214  [n1,s1,r11/1:/{Table/14-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:07.198098 148992 storage/replica_command.go:282  [split,n1,s1,r12/1:/{Table/15-Max}] initiating a split of this range at key /Table/16 [r13]
I180817 07:08:07.489087 148416 server/server.go:1592  [n1] done ensuring all necessary migrations have run
I180817 07:08:07.489394 148416 server/server.go:1595  [n1] serving sql connections
I180817 07:08:08.511917 148896 server/server_update.go:67  [n1] no need to upgrade, cluster already at the newest version
I180817 07:08:08.979803 148726 server/status/runtime.go:433  [n1] runtime stats: 1.3 GiB RSS, 268 goroutines, 23 MiB/101 MiB/146 MiB GO alloc/idle/total, 40 MiB/85 MiB CGO alloc/total, 389.31cgo/sec, 0.25/0.05 %(u/s)time, 0.01 %gc (1x)
I180817 07:08:09.013499 149059 sql/event_log.go:126  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:42079} Attrs: Locality: ServerVersion:2.0-11 BuildTag:v2.1.0-alpha.20180702-1701-gffb539f StartedAt:1534489657969557679 LocalityAddress:[]} ClusterID:bd32192e-e9d2-439d-be56-5148d23940b9 StartedAt:1534489657969557679 LastUp:1534489657969557679}
I180817 07:08:09.388499 148673 storage/replica_proposal.go:214  [n1,s1,r12/1:/{Table/15-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:09.447512 149049 storage/replica_command.go:282  [split,n1,s1,r13/1:/{Table/16-Max}] initiating a split of this range at key /Table/17 [r14]
I180817 07:08:10.109825 148728 server/status/recorder.go:652  [n1,summaries] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180817 07:08:10.216174 148695 storage/replica_proposal.go:214  [n1,s1,r13/1:/{Table/16-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:10.231979 149053 storage/replica_command.go:282  [split,n1,s1,r14/1:/{Table/17-Max}] initiating a split of this range at key /Table/18 [r15]
I180817 07:08:11.831340 148650 storage/replica_proposal.go:214  [n1,s1,r14/1:/{Table/17-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:11.846058 149094 storage/replica_command.go:282  [split,n1,s1,r15/1:/{Table/18-Max}] initiating a split of this range at key /Table/19 [r16]
I180817 07:08:13.528318 148683 storage/replica_proposal.go:214  [n1,s1,r15/1:/{Table/18-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:13.612104 149042 storage/replica_command.go:282  [split,n1,s1,r16/1:/{Table/19-Max}] initiating a split of this range at key /Table/20 [r17]
I180817 07:08:15.465023 148521 storage/replica_proposal.go:214  [n1,s1,r16/1:/{Table/19-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:15.519913 149057 storage/replica_command.go:282  [split,n1,s1,r17/1:/{Table/20-Max}] initiating a split of this range at key /Table/21 [r18]
I180817 07:08:16.799160 148672 storage/replica_proposal.go:214  [n1,s1,r17/1:/{Table/20-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:16.899070 149100 storage/replica_command.go:282  [split,n1,s1,r18/1:/{Table/21-Max}] initiating a split of this range at key /Table/22 [r19]
I180817 07:08:17.660847 148691 storage/replica_proposal.go:214  [n1,s1,r18/1:/{Table/21-Max}] new range lease repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0 following repl=(n1,s1):1 seq=3 start=1534489658.135072753,0 epo=1 pro=1534489676.413415034,0
I180817 07:08:17.711200 149105 storage/replica_command.go:282  [split,n1,s1,r19/1:/{Table/22-Max}] initiating a split of this range at key /Table/23 [r20]
I180817 07:08:18.272324 148416 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      [async] transport racer
1      [async] storage.split: processing replica
1      [async] kv.TxnCoordSender: heartbeat loop
I180817 07:08:18.273726 148635 kv/transport_race.go:67  transport race promotion: ran 44 iterations on up to 708 requests
I180817 07:08:18.274194 148416 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      [async] storage.split: processing replica
1      [async] kv.TxnCoordSender: heartbeat loop
I180817 07:08:18.277952 148416 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
1      [async] storage.split: processing replica
W180817 07:08:18.409664 149105 internal/client/txn.go:514  [split,n1,s1,r19/1:/{Table/22-Max}] failure aborting transaction: node unavailable; try another peer; abort caused by: log-range-event: node unavailable; try another peer
E180817 07:08:18.410357 149105 storage/queue.go:788  [split,n1,s1,r19/1:/{Table/22-Max}] unable to split [n1,s1,r19/1:/{Table/22-Max}] at key "/Table/23": split at key /Table/23 failed: log-range-event: node unavailable; try another peer
I180817 07:08:18.410745 148416 util/stop/stopper.go:537  quiescing; tasks left:
1      ts.poller: poll
I180817 07:08:18.968846 148726 server/status/runtime.go:433  [n1] runtime stats: 1.3 GiB RSS, 230 goroutines, 18 MiB/106 MiB/146 MiB GO alloc/idle/total, 38 MiB/85 MiB CGO alloc/total, 370.32cgo/sec, 0.26/0.02 %(u/s)time, 0.02 %gc (2x)
W180817 07:08:20.101497 148728 server/node.go:857  [n1,summaries] error recording status summaries: node unavailable; try another peer
W180817 07:08:22.560970 148727 ts/db.go:195  [n1,ts-poll] error writing time series data: node unavailable; try another peer
--- FAIL: TestSingleKey (48.71s)
	test_server_shim.go:176: had 18 ranges at startup, expected 22

```

--- a/pkg/storage/replicate_queue.go
+++ b/pkg/storage/replicate_queue.go
@@ -86,6 +86,25 @@ var (
 	}
 )
 
+// quorumError indicates a retryable error condition which sends replicas being
+// processed through the replicate queue into purgatory so that they can be
+// retried quickly as soon as nodes come online.
+type quorumError struct {
+	msg string
+}
+
+func newQuorumError(f string, args ...interface{}) *quorumError {
+	return &quorumError{
+		msg: fmt.Sprintf(f, args...),
+	}
+}
+
+func (e *quorumError) Error() string {
+	return e.msg
+}
+
+func (*quorumError) purgatoryErrorMarker() {}
+
 // ReplicateQueueMetrics is the set of metrics for the replicate queue.
 type ReplicateQueueMetrics struct {
 	AddReplicaCount        *metric.Counter
@@ -267,7 +286,7 @@ func (rq *replicateQueue) processOneChange(
 	{
 		quorum := computeQuorum(len(desc.Replicas))
 		if lr := len(liveReplicas); lr < quorum {
-			return false, errors.Errorf(
+			return false, newQuorumError(
 				"range requires a replication change, but lacks a quorum of live replicas (%d/%d)", lr, quorum)
 		}
 	}

